### ISSUE TITLE
How to ensure that subchart is installed before the parent chart

### ISSUE THREAD

**Issue by innerpeacez** [helm/helm#6737]
The parent chart depends on a subchart. I want to ensure that the subchart is installed first when executing `helm install`, and then install the parent chart. What should I do? I need help, please.

Output of `helm version`:
`version.BuildInfo{Version:"v3.0.0-beta.4", GitCommit:"7ffc879f137bd3a69eea53349b01f05e3d1d2385", GitTreeState:"dirty", GoVersion:"go1.13.1"}`
Output of `kubectl version`:
`Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.1", GitCommit:"eec55b9ba98609a46fee712359c7b5b365bdd920", GitTreeState:"clean", BuildDate:"2018-12-13T10:39:04Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.8", GitCommit:"211047e9a1922595eaa3a1127ed365e9299a6c23", GitTreeState:"clean", BuildDate:"2019-10-15T12:02:12Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}`
Cloud Provider/Platform (AKS, GKE, Minikube etc.):

**Comment by bridgetkromhout** [comment]
Hi, @innerpeacez - you probably want to look at Chart Hooks: https://helm.sh/docs/topics/charts_hooks/#hooks-and-the-release-lifecycle (and note that it explains: "When subcharts declare hooks, those are also evaluated. There is no way for a top-level chart to disable the hooks declared by subcharts.")

I'll resolve this issue now, but please let us know if you have further questions. Thanks!

**Comment by aneelmonangi** [comment]
Any update on this issue?
When the subchart is having pre-upgrade hook, and mail chart is upgrading enabling the same subchart. In this case pre-upgrade hook is executing, but by this time the subchart was not installed.
How to overcome this issue.
Please comment, if anything wrong here.

Thanks,
Aneel

**Comment by AmarendraSingh88** [comment]
We have a similar issue where we want to wait for some time (for the creation of secrets etc.) before running a sub chart. We have one umbrella chart and added the pre-install hook in the sub-chart. As soon as the chart is deployed, the pre-install hook gets triggered before starting any sub chart, which is not helping, as the secrets get created in another sub chart.
How can we execute the pre-install hook before installing one sub-chart while at the same time other sub-charts should get installed?

**Comment by patsevanton** [comment]
Somebody have example ensure that subchart is installed before the parent chart ? Thanks!

**Comment by Phil1602** [comment]
We are facing similar issues.

## how to reproduce
- Umbrella chart, where parent charts `templates` folder contains a resource named `my-foo` of `Kind: Foo`
- sub-chart having a resource `Foo` of type `CustomResourceDefinition` in `templates` folder

## observation
the installation/upgrade fails with the following reason:

```
Error: UPGRADE FAILED: unable to build kubernetes objects from current release manifest: resource mapping not found for name: "my-foo" namespace: "<REDACTED>" from "": no matches for kind "Foo" in version "foo/v1"
ensure CRDs are installed first
```

The same issue also persists when I add the following hook to `my-foo` resource:

```yaml
helm.sh/hook: post-install,post-upgrade
```

It seems that chart order has precedence over resource order, so that all resources from the parent chart are applied before any of the sub-chart resources is, even though they are postponed using the hook annotations. 

Is this correct and expected?

## expected behavior
At least in my example (depending on CRDs from subchart), the [defined ordering](https://helm.sh/docs/intro/using_helm/#helm-install-installing-a-package) of helm would probably already fix this issue generally for almost all cases  - if this ordering would be applied to all resourced independent if they are part of the parent or any sub-chart.


Any ideas about this @bridgetkromhout? Is the ordering we are seeing here (chart ordering has precedence over resource kinds) correct?
